### ISSUE TITLE
 Feature Request: Better control of train and evaluation modes

### ISSUE THREAD

**Issue by kefirski** [tensorflow/tensorflow#26552]
**System information**
- TensorFlow version (you are using): tf 2.0 alpha
- Are you willing to contribute it (Yes/No): Yes

Current approach for handling train and evaluation modes involves passing additional argument to `call` method like this

```
def call(self, inputs, training=None):
    ...
```

If we would consider that we have some model with deep nested layers that contain `Dropout`, `Batch Norm` and other layers, then it would be necessary to pass `training` argument to all of them through several calls.

This approach have several drawbacks.

First of all, it could be potential bug in the model, since programmer could forget that propagation of some layer depends on training state.
Also, passing `training` argument is extremely verbose.

At the same time, PyTorch solves this task in much more elegant way – every layer have additional state `train: bool`. Since we are aware of nested layers for some model, setting `training = False` on topmost layer will set all children layers to this mode without being necessary to explicitly handle passing training state argument.

Implementation of this feature will not change current API, but will add new methods/functionality to TensorFlow. It could be two additional methods `Model.train()` and `Model.eval()` that recursively change every layer state, which affects on `call` method like `training` argument do.

This feature will simplify creation of custom models for both production and research tasks. It will make code much more clear and will make TensorFlow more attractive for active PyTorch users.

**Comment by Oktai15** [comment]
@kefirski great idea, support you, man!

**Comment by shashvatshahi1998** [comment]
i am willing to work on this issue please anybody guide me a little bit.

**Comment by karmel** [comment]
@kefirski I'm assuming you are talking about Keras models specifically. If so, can you provide some sample code that shows the expected outcome? I don't entirely follow your solution here. CC @tomerk , FYI.

**Comment by Oktai15** [comment]
@karmel, hi! @kefirski said about opportunity to change mode (training or evaluation) of model in one method like in PyTorch. See source code [here](https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#Module.train) or attached screenshot. 
![image](https://user-images.githubusercontent.com/17337773/59145317-d0350900-89ea-11e9-9f0f-c5695016463b.png)

It's really convenient, because you don't need to push `training` variable in every module that I create.

**Comment by karmel** [comment]
@omalleyt12 as you've been thinking in this direction recently.

**Comment by omalleyt12** [comment]
Yep @kefirski I think we've been thinking in similar directions :) Models like this now work:

```
class MyModel(tf.keras.Model):
  def __init__(self):
    super(MyModel, self).__init__()
    self.bn = tf.keras.layers.BatchNormalization()
    self.dropout = tf.keras.layers.Dropout()

  def call(self, inputs):
   x = self.dropout(inputs)
   return self.bn(x)

# `training` value passed to a top-level layer is used as the default for sublayers
model = MyModel()
model(x, training=True) # BN and Dropout run in training mode
model(x, training=False) # BN and Dropout run in inference mode
```
That is, the `training` arg in `__call__` is still kept as the way to set training/inference mode on a Layer/Model (so as not to introduce too many different ways to handle this), but propagation of this arg is automatically handled for you (unless you explicitly pass in `training` to sublayers). Thus `model(..., training=True)` is the equivalent to what you are suggesting via `model.train()`.

The order of preference for how the `training` arg evaluates in a sublayer is:

1) The value explicitly passed to the layer
2) The value passed to the layer's caller
3) The learning_phase if set via `tf.keras.backend.set_learning_phase`
4) The default value of  `training` in the Layer's signature

**Comment by kefirski** [comment]
@omalleyt12 yep, it seems to be a nice way to handle this issue.

I guess, the implementation of this feature would imply that when we do `layer(..., training=True)`, we have to change some state variables of learning phase for `layer` sublayers right before calling `layer.call` method. In this case we would obtain order of preference that you have described above.

**Comment by Oktai15** [comment]
Why did you close it, @kefirski? This feature has been already done?

**Comment by omalleyt12** [comment]
@Oktai15 yes this feature is now implemented in the latest 2.0 preview :)
